text/0039-service-overhaul.md

```diff
@@ -0,0 +1,462 @@
+    - `"x264opts"` (optional): default option for x264 for this service.
+    - `"max video bitrate"` (optional): maximum video bitrate for this service.
+    - `"max audio bitrate"` (optional): maximum audio bitrate for this service.
+    - `"max fps"` (optional): maximum frame per second for this service.
+    - `"supported resolutions"` (optional): array of supported resolutions for this service.
+    - `"bitrate matrix"` (optional): array of maximum bitrate for a certain resolution/FPS combo for this service.
+      - `"res"`: resolution.
+      - `"fps"`: frame per seconds.
+      - `"max bitrate"`: maximum video bitrate.
+
+##### Issues with this format
+- The naming scheme is a mix of object name with space and some with underscores.
+- About `"output"`, OBS consider every service as RTMP if not added and it's not a recommended settings at all. It makes OBS use the right protocol for the service. Also this prevent a service of being multi-protocol.
+- About `"recommended"`, most of the options seems to H264 related
+- `"common"`, the name makes it not understandable at the first sight maybe adding some documention would be good thing.
+
+#### New format (WIP)
+The new format will be parsable with a new interface library named `service-json-parser` to allow first-party plugins to read this format rather than recreating the wheel.
+
+The new format will be considered as the version 4.
+
+This format allow some specific additions parsed by getter added to the concerned plugin.
+
+Multi-service plugins should have a JSON root that look like this:
+```json
+{
+    "format_version": 4,
+    "services": []
+}
+```
+`"services"` is an array of service objects.
+
+Service plugin like integrations one should have a JSON root that look like this:
+```json
+{
+    "format_version": 4,
+    "service": {}
+}
+```
+`"service"` is the service object itself.
+
+##### Plugins specific additions
+Those additions are added like extensions to the service object format.
+
+Getters for those customisations are implemented in the plugin itself.
+- In `obs-services`:
+  - `"id"` (required): service identifier, services are no longer identified by their name. Added because this plugin is multi-services.
+  - `"name"` (required): name of the service. Added because this plugin is multi-services.
+  - `"common"` (optional, false by default): this object is added in service objects of this plugin to identify which service in "common" or not. No third-party plugin should have this behavior.
+- In `obs-youtube`:
+  - Servers name are translated.
+  - `"name_suffix"` in `"servers"` (optional): suffix added to the server name after being translated.
+
+Specific documentations about adding service in `obs-services` should be made.
+
+##### Service object format
+<sup>S</sup>: Should be required by any first-party plugin
+
+<sup>O</sup>: Optional in `obs-services`, can be required or not used at all in first-party one-service plugins depending on their implementations.
```

Comment:
Due to the size & complexity of the available options here, it would be worth [JSON Schema](https://json-schema.org/) document.

Reply:
Didn't know about this, I never used to heavily JSON. I will check this link.